ecdsa/ta/ecdsa_ta.c

```diff
@@ -0,0 +1,342 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
+ */
+
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <ecdsa_ta.h>
+
+struct ecdsa {
+	TEE_ObjectHandle keypair;
+	TEE_ObjectHandle public_key;
+	uint32_t algo_id;
+	uint32_t hash_algo;
+	uint32_t key_sz;
+};
+
+TEE_Result TA_CreateEntryPoint(void)
+{
+	DMSG("TA created");
+	return TEE_SUCCESS;
+}
+
+void TA_DestroyEntryPoint(void)
+{
+	/* TA destroyed */
+}
+
+TEE_Result TA_OpenSessionEntryPoint(uint32_t param_types,
+				    TEE_Param params[4],
+				    void **sess_ctx)
+{
+	struct ecdsa *sess;
+
+	sess = TEE_Malloc(sizeof(*sess), 0);
+	if (!sess)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	sess->keypair = TEE_HANDLE_NULL;
+	sess->public_key = TEE_HANDLE_NULL;
+
+	(void)param_types;
+	(void)params;
+
+	*sess_ctx = (void *)sess;
+	DMSG("Session %p: newly allocated", *sess_ctx);
+	return TEE_SUCCESS;
+
+}
+
+void TA_CloseSessionEntryPoint(void *sess_ctx)
+{
+	struct ecdsa *sess;
+
+	sess = (struct ecdsa *)sess_ctx;
+
+	/* release session */
+	if (sess->keypair != TEE_HANDLE_NULL)
+		TEE_FreeTransientObject(sess->keypair);
+	if (sess->public_key != TEE_HANDLE_NULL)
+		TEE_FreeTransientObject(sess->public_key);
+
+	TEE_Free(sess);
+}
+
+static TEE_Result get_hash_algo(uint32_t algo, uint32_t *hash_algo)
+{
+	switch (algo) {
+	case TEE_ALG_ECDSA_SHA1:
+		*hash_algo = TEE_ALG_SHA1;
+		return TEE_SUCCESS;
+	case TEE_ALG_ECDSA_SHA224:
+		*hash_algo = TEE_ALG_SHA224;
+		return TEE_SUCCESS;
+	case TEE_ALG_ECDSA_SHA256:
+		*hash_algo = TEE_ALG_SHA256;
+		return TEE_SUCCESS;
+	case TEE_ALG_ECDSA_SHA384:
+		*hash_algo = TEE_ALG_SHA384;
+		return TEE_SUCCESS;
+	case TEE_ALG_ECDSA_SHA512:
+		*hash_algo = TEE_ALG_SHA512;
+		return TEE_SUCCESS;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+}
+
+static TEE_Result ta2tee_algo_id(uint32_t param, uint32_t *algo_id)
+{
+	switch (param) {
+	case TA_ALG_ECDSA_SHA1:
+		*algo_id = TEE_ALG_ECDSA_SHA1;
```

Comment:
I'd prefer the client to provide this value directly to avoid translating it again. The algorithm values are defined in a GP standard, so they are "well known" and shouldn't change.